File: clients/java/src/main/java/spb_ai_champ/model/Resource.java
package spb_ai_champ.model;

import spb_ai_champ.util.StreamUtil;

/**
 * TODO - Document
 */
public enum Resource {
    /**
     * TODO - Document
     */
    STONE(0),
    /**
     * TODO - Document
     */
    ORE(1),
    /**
     * TODO - Document
     */
    SAND(2),
    /**
     * TODO - Document
     */
    ORGANICS(3),
    /**
     * TODO - Document
     */
    METAL(4),
    /**
     * TODO - Document
     */
    SILICON(5),
    /**
     * TODO - Document
     */
    PLASTIC(6),
    /**
     * TODO - Document
     */
    CHIP(7),
    /**
     * TODO - Document
     */
    ACCUMULATOR(8);

    public int tag;

    Resource(int tag) {
        this.tag = tag;
    }

    /**
     * Read Resource from input stream
     */
    public static Resource readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
        case 0:
            return STONE;
        case 1:
            return ORE;
        case 2:
            return SAND;
        case 3:
            return ORGANICS;
        case 4:
            return METAL;
        case 5:
            return SILICON;
        case 6:
            return PLASTIC;
        case 7:
            return CHIP;
        case 8:
            return ACCUMULATOR;
        default:
            throw new java.io.IOException("Unexpected tag value");
        }
    }
}

File: clients/kotlin/src/main/kotlin/spb_ai_champ/model/Resource.kt
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * Resource type
 */
enum class Resource private constructor(val tag: Int) {
    /**
     * Stone
     */
    STONE(0),
    /**
     * Ore
     */
    ORE(1),
    /**
     * Sand
     */
    SAND(2),
    /**
     * Organics
     */
    ORGANICS(3),
    /**
     * Metal
     */
    METAL(4),
    /**
     * Silicon
     */
    SILICON(5),
    /**
     * Plastic
     */
    PLASTIC(6),
    /**
     * Chip
     */
    CHIP(7),
    /**
     * Accumulator
     */
    ACCUMULATOR(8);

    companion object {
        /**
         * Read Resource from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Resource {
            return when (StreamUtil.readInt(stream)) {
            STONE.tag -> STONE
            ORE.tag -> ORE
            SAND.tag -> SAND
            ORGANICS.tag -> ORGANICS
            METAL.tag -> METAL
            SILICON.tag -> SILICON
            PLASTIC.tag -> PLASTIC
            CHIP.tag -> CHIP
            ACCUMULATOR.tag -> ACCUMULATOR
            else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }
}